src/SearchQueryService/Controllers/SearchController.cs

```diff
@@ -199,28 +220,87 @@ private IEnumerable<SolrDelete> ConvertAzDocsForDelete(AzPost azDocs)
+                }
+                else
+                {
+                    for (int i = 0; i < item.Value.Count; i++)
+                    {
+                        item.Value[i] = ConvertValue(item.Value[i]);
+                    }
+                    converted[item.Key] = JsonSerializer.SerializeToElement(new SolrSetProperty(item.Value));
+                }
+            }
+
+            return converted;
+        }
+
+        private static Dictionary<string, List<JsonElement>> Flatten(Dictionary<string, JsonElement> json)
         {
-            var convertedDocument = new Dictionary<string, object>();
-            foreach (var kv in document)
+            var flattened = new Dictionary<string, List<JsonElement>>();
+            foreach (var kv in json)
             {
-                if (string.Equals(kv.Key, "id", StringComparison.OrdinalIgnoreCase))
+                if (kv.Value.ValueKind == JsonValueKind.Array)
                 {
-                    convertedDocument["id"] = kv.Value;
+                    foreach (var arrayItem in kv.Value.EnumerateArray())
+                    {
+                        foreach (var property in arrayItem.EnumerateObject())
+                        {
+                            string propName = kv.Key + "." + property.Name;
+                            if (flattened.ContainsKey(propName))
+                            {
+                                flattened[propName].Add(property.Value);
+                            }
+                            else
+                            {
+                                flattened.Add(propName, new List<JsonElement>() { property.Value });
+                            }
+                        }
+                    }
                 }
                 else
                 {
-                    convertedDocument[kv.Key] = ConvertValues(kv.Value);
+                    flattened.Add(kv.Key, new List<JsonElement>() { kv.Value });
                 }
             }
 
-            return convertedDocument;
+            return flattened;
+        }
+
+        private List<Dictionary<string, JsonElement>> ConvertAzDocsForAdd(AzPost azDocs)
+        {
+            foreach (var doc in azDocs.Value)
+            {
+                FixPrimaryKeyForAdd(doc);
+
+                foreach (var prop in doc)
+                {
+                    doc[prop.Key] = ConvertValue(prop.Value);
+                }
+            }
+
+            return azDocs.Value;
+        }
+
+        private static void FixPrimaryKeyForAdd(Dictionary<string, JsonElement> document)
+        {
+            if (document.ContainsKey("Id"))
+            {
+                JsonElement id = document["Id"];
+                document.Remove("Id");
+                document["id"] = id;
+            }
         }
 
-        private dynamic ConvertValues(JsonElement value)
+        private JsonElement ConvertValue(JsonElement value)
```

Comment:
This could probably be moved somewhere else, along with the `_valueReplacements`, but i couldn't figure out where.